### PR TITLE
Add loading tags, starting with croissant and webdataset

### PR DESCRIPTION
**Pull request by lhoestq**
Added two tags:
- "croissant" for datasets that have a valid info (= a valid croissant json)
- "webdataset" fro datasets with `info["builder_name"] == "webdataset"`

I implemented this using a new job "dataset-loading-tags" that aims at detecting tags related to dataset loading tools, and then adding the tags to "dataset-hub-cache".

Later we can enrich "dataset-loading-tags" with loading argument that can be used when showing code snippets in dataset pages.

TODO
- [x] add to hub-cache
- [x] tests

### CODE REVIEW

File: services/worker/src/worker/dtos.py [hunk 1]
```diff
@@ -222,6 +222,7 @@ class DatasetHubCacheResponse(TypedDict):
     viewer: bool
     partial: bool
     num_rows: int
+    tags: list[str]
```
Comment (severo) [inline services/worker/src/worker/dtos.py:225]: let's use `Literal["croissant", "webdataset"]` (let's create a dedicate type for that)

File: services/worker/src/worker/dtos.py [hunk 2]
```diff
@@ -250,3 +251,7 @@ class DatasetSizeResponse(TypedDict):
     pending: list[PreviousJob]
     failed: list[PreviousJob]
     partial: bool
+
+
+class DatasetLoadingTagsResponse(TypedDict):
+    tags: list[str]
```
Comment (severo) [inline services/worker/src/worker/dtos.py:257]: same

File: services/worker/src/worker/job_runners/dataset/hub_cache.py [hunk 3]
```diff
@@ -63,12 +63,24 @@ def compute_hub_cache_response(dataset: str) -> tuple[DatasetHubCacheResponse, f
 
     progress = min((p for p in [is_valid_progress, size_progress] if p is not None), default=0.0)
 
+    tags: list[str] = []
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/hub_cache.py:66]: same

File: services/worker/src/worker/job_runners/dataset/loading_tags.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    get_previous_step_or_raise,
+)
+
+from worker.dtos import (
+    CompleteJobResult,
+    DatasetLoadingTagsResponse,
+)
+from worker.job_runners.dataset.dataset_job_runner import DatasetJobRunner
+
+MAX_CONFIGS = 100
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/loading_tags.py:19]: let's move to libcommon.constants (or reuse an existing constant there)?

File: services/worker/src/worker/job_runners/dataset/loading_tags.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    Raises:
+        [~`libcommon.simple_cache.CachedArtifactError`]:
+          If the previous step gave an error.
+        [~`libcommon.exceptions.PreviousStepFormatError`]:
+            If the content of the previous step has not the expected format
+
+    Returns:
+        `DatasetLoadingTagsResponse`: The dataset_loading_tags_response (list of tags).
+    """
+    logging.info(f"get 'dataset-info' for {dataset=}")
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/loading_tags.py:39]:
```suggestion
    logging.info(f"get 'dataset-loading-tags' for {dataset=}")
```

File: services/worker/src/worker/job_runners/dataset/loading_tags.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            A namespace (user or an organization) and a repo name separated by a `/`.
+
+    Raises:
+        [~`libcommon.simple_cache.CachedArtifactError`]:
+          If the previous step gave an error.
+        [~`libcommon.exceptions.PreviousStepFormatError`]:
+            If the content of the previous step has not the expected format
+
+    Returns:
+        `DatasetLoadingTagsResponse`: The dataset_loading_tags_response (list of tags).
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/loading_tags.py:37]:
```suggestion
        `DatasetLoadingTagsResponse`: The dataset-loading-tags response (list of tags).
```

File: services/worker/src/worker/job_runners/dataset/loading_tags.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    logging.info(f"get 'dataset-info' for {dataset=}")
+
+    dataset_info_best_response = get_previous_step_or_raise(kinds=["dataset-info"], dataset=dataset)
+    http_status = dataset_info_best_response.response["http_status"]
+    tags: list[str] = []
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/loading_tags.py:43]: same

**Review by severo — COMMENTED** [review]
also: should we add a migration step to add `tags: []` to all the responses? or just wait for the backfill to update all the `hub-cache` entries?

**Comment by lhoestq** [comment]
Hmm maybe a migration would be faster indeed, and no need to increase the job version in this case

File: jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+# connection already occurred in the main.py (caveat: we use globals)
+class MigrationAddTagsToHubCacheCacheResponse(Migration):
+    def up(self) -> None:
+        # See https://docs.mongoengine.org/guide/migration.html#example-1-addition-of-a-field
+        db = get_db(CACHE_MONGOENGINE_ALIAS)
+        logging.info(
+            "If missing, add the 'tags' field with the default value ([]) to the success cached results of dataset-hub-cache"
+        )
+        db[CACHE_COLLECTION_RESPONSES].update_many(
+            {"kind": "dataset-hub-cache", "http_status": 200, "tags": {"$exists": False}}, {"$set": {"tags": []}}
```
Comment (severo) [inline jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py:23]: we want to add this in "content.tags", not to the root

Comment (lhoestq) [reply jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py:23]: good catch

File: jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+class MigrationAddTagsToHubCacheCacheResponse(Migration):
+    def up(self) -> None:
+        # See https://docs.mongoengine.org/guide/migration.html#example-1-addition-of-a-field
+        db = get_db(CACHE_MONGOENGINE_ALIAS)
+        logging.info(
+            "If missing, add the 'tags' field with the default value ([]) to the success cached results of dataset-hub-cache"
+        )
+        db[CACHE_COLLECTION_RESPONSES].update_many(
+            {"kind": "dataset-hub-cache", "http_status": 200, "content.tags": {"$exists": False}},
+            {"$set": {"tags": []}},
```
Comment (severo) [inline jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py:24]: no :)

```suggestion
            {"$set": {"content.tags": []}},
```

Comment (lhoestq) [reply jobs/mongodb_migration/src/mongodb_migration/migrations/_20240206153000_cache_add_tags_in_hub_cache.py]: 😬

**Comment by lhoestq** [comment]
alright all green now, lmk if it's good for you

**Review by severo — APPROVED** [review]

**Pull request merged**